storage/block/block.go

```diff
@@ -261,76 +271,85 @@ func (b *dbBlock) Merge(other DatabaseBlock) error {
 	return nil
 }
 
-func (b *dbBlock) Reset(start time.Time, segment ts.Segment) {
+func (b *dbBlock) Reset(start time.Time, blockSize time.Duration, segment ts.Segment) {
 	b.Lock()
 	defer b.Unlock()
-	b.resetNewBlockStartWithLock(start)
+	b.resetNewBlockStartWithLock(start, blockSize)
 	b.resetSegmentWithLock(segment)
 }
 
 func (b *dbBlock) ResetRetrievable(
 	start time.Time,
+	blockSize time.Duration,
 	retriever DatabaseShardBlockRetriever,
 	metadata RetrievableBlockMetadata,
 ) {
 	b.Lock()
 	defer b.Unlock()
-	b.resetNewBlockStartWithLock(start)
+	b.resetNewBlockStartWithLock(start, blockSize)
 	b.resetRetrievableWithLock(retriever, metadata)
 }
 
-func (b *dbBlock) stream(ctx context.Context) (xio.SegmentReader, error) {
+func (b *dbBlock) stream(ctx context.Context) (xio.BlockReader, error) {
 	b.ctx.DependsOn(ctx)
 
+	start := b.startWithLock()
+	end := start.Add(b.blockSize)
+
 	// If the block retrieve ID is set then it must be retrieved
 	var (
-		stream xio.SegmentReader
-		err    error
+		blockReader xio.BlockReader
+		err         error
 	)
 	if b.retriever != nil {
-		start := b.startWithLock()
-		stream, err = b.retriever.Stream(ctx, b.retrieveID, start, b)
+		blockReader, err = b.retriever.Stream(ctx, b.retrieveID, start, end, b)
 		if err != nil {
-			return nil, err
+			return xio.EmptyBlockReader, err
 		}
 	} else {
-		stream = b.opts.SegmentReaderPool().Get()
-		stream.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
-		ctx.RegisterFinalizer(stream)
+		blockReader = xio.BlockReader{
+			SegmentReader: b.opts.SegmentReaderPool().Get(),
+			Start:         start,
+			End:           end,
+		}
+		blockReader.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
+		ctx.RegisterFinalizer(blockReader)
```

Comment:
Hm, maybe just finalize the segment reader here?  Otherwise the `blockReader` will need to be coerced into an interface type here too (and will end up being heap allocated).  

However all we really need to do is finalize the segment reader and not the encapsulating block reader itself.